File: JavaScript/005-smallest-multiple.js
/*
	2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

	What is the smallest positive number that is evenly divisible by all of the numbers from 1 to n?
*/

function smallestMultiple(n) {
	
	// The eventual result
	let currentNum = 1;

	// Keep going until the result is found and returned
	while(true) {

		// Only try dividing if the current number is greater than or equal to the argument, as it must also be divisible by it
		if(currentNum >= n) {

			// If the number couldn't be divisible by any number until the argument, then return it
			if(isDivisible(currentNum, n)) {
				return currentNum;
			}

		}

		currentNum++;
	}

}

function isDivisible(number, n) {

	for(let i = 1; i <= n; i++) {

		// If the number could be divisible by any number, then it's not the correct number, we break and don't have to check other numbers
		if(number % i !== 0) {
			return false;
		}

	}

	return true;
}

let result = smallestMultiple(10);
console.log(result);

File: Ruby/005-smallest-multiple.rb
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

def smallestMultiple(maxFactor) # maxFactor - max positive factor that use to check the divisible from 1 to maxFactor
  lcm = 1 # least common multiple - the smallest positive integer that is divisible by two integers
  (2..maxFactor).each do |factor| # Iterate from 2 to maxFactor. No need to consider 1 here since it's a factor of all integers.
    lcm *= factor / greatestCommonDivisor(lcm, factor)
  end

  lcm
end

# Find the largest positive integer that divides each of the integers
# Ex:
#   <factor>  -->  <greatestCommonDivisor>
#       2     -->     1
#       3     -->     1
#       4     -->     2
#       5     -->     1
#       6     -->     6
#       7     -->     1
#       8     -->     4
#       ...............
#       ...............
def greatestCommonDivisor(lcm, factor)
  while factor > 0
    lcm %= factor
    return factor if lcm == 0
    factor %= lcm
  end

  lcm
end

puts smallestMultiple(20) # put 20 here to solve problem 5
